File: day01/main.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <stdbool.h>

int turnDial(int dial, int turn, char direction, int *clicks) {
    if (direction == 'R') {
        *clicks += (dial + turn) / 100;
        return (dial + turn) % 100;
    } else if (direction == 'L') {
        *clicks += (dial - turn) / -100 + (turn >= dial) - (dial == 0);
        return (dial + 100 - (turn % 100)) % 100;
    }

    fprintf(stderr, "bad direction!");
    return -1;
}

int main(int argc, char* argv[]) {
    char *path = argc > 1 ? argv[1] : "input.txt"; 
    FILE *f = fopen(path, "r");

    int dial = 50;

    char l;
    int turn, countZ = 0, countClick = 0;
    while (fscanf(f, "%c%d\n", &l, &turn) != EOF) {
        dial = turnDial(dial, turn, l, &countClick);
        dial = (dial + (turn / 100 + 1) * 100) % 100;
        if (dial == 0) countZ++;
    }

    printf("count zero: %d\n", countZ);
    printf("count click: %d\n", countClick);
    fclose(f);
}

File: day01/src/main.zig
const std = @import("std");

fn turnDial(dial: i32, turn: i32, direction: u8, clicks: *u32) !i32 {
    if (direction == 'R') {
        clicks.* += @intCast(@divTrunc(dial + turn, 100));
        return @mod(dial + turn, 100);
    } else if (direction == 'L') {
        clicks.* += @intCast(@divTrunc(dial - turn, -100) + @intFromBool(turn >= dial) - @intFromBool(dial == 0));
        return @mod(dial + 100 - @mod(turn, 100), 100);
    }

    return error.BadDirection;
}

fn openInput() !std.fs.File {
    var args = std.process.args();
    _ = args.skip();
    const path = args.next() orelse "input.txt";
    return std.fs.cwd().openFile(path, .{});
}

pub fn main() !void {
    const input = try openInput();
    defer input.close();
    var buf: [512]u8 = undefined;
    var reader = input.reader(&buf);
    const interface = &reader.interface;

    var dial: i32 = 50;
    var countZ: u32 = 0;
    var countClicks: u32 = 0;

    while (try interface.takeDelimiter('\n')) |line| {
        const dir = line[0];
        const turn = try std.fmt.parseInt(i32, line[1..], 10);
        dial = try turnDial(dial, turn, dir, &countClicks);
        if (dial == 0) countZ += 1;
    }

    var stdout_wrt = std.fs.File.stdout().writer(&buf);
    const stdout = &stdout_wrt.interface;
    defer stdout.flush() catch {};

    try stdout.print("count zero: {}\ncount click: {}\n", .{ countZ, countClicks });
}
